runtime/rococo/src/propose_parachain.rs

```diff
@@ -0,0 +1,836 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A pallet for proposing a parachain for Rococo.
+
+use frame_support::{
+	decl_event, decl_error, decl_module, traits::{Get, ReservableCurrency, EnsureOrigin, Currency},
+	decl_storage, ensure, IterableStorageMap,
+};
+use primitives::v0::{Id as ParaId, Info as ParaInfo, Scheduling, HeadData, ValidationCode};
+use polkadot_parachain::primitives::AccountIdConversion;
+use system::{ensure_signed, EnsureOneOf, EnsureSigned};
+use sp_runtime::{Either, traits::BadOrigin};
+use sp_staking::SessionIndex;
+use sp_std::vec::Vec;
+use runtime_common::registrar::Registrar;
+
+type EnsurePriviledgedOrSigned<T> = EnsureOneOf<
+	<T as system::Trait>::AccountId,
+	<T as Trait>::PriviledgedOrigin,
+	EnsureSigned<<T as system::Trait>::AccountId>
+>;
+
+type Session<T> = session::Module<T>;
+
+type BalanceOf<T> = <<T as runtime_common::registrar::Trait>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+/// Configuration for the parachain proposer.
+pub trait Trait: session::Trait + runtime_common::parachains::Trait + runtime_common::registrar::Trait {
+	/// The overreaching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+
+	/// The maximum name length of a parachain.
+	type MaxNameLength: Get<u32>;
+
+	/// The amount that should be deposited when creating a proposal.
+	type ProposeDeposit: Get<BalanceOf<Self>>;
+
+	/// Priviledged origin that can approve/cancel/deregister parachain and proposals.
+	type PriviledgedOrigin: EnsureOrigin<<Self as system::Trait>::Origin, Success = ()>;
+}
+
+/// A proposal for adding a parachain to the relay chain.
+#[derive(codec::Encode, codec::Decode)]
+struct Proposal<AccountId, ValidatorId, Balance> {
+	/// The account that proposed this parachain.
+	proposer: AccountId,
+	/// The validation WASM code of the parachain.
+	validation_function: ValidationCode,
```

Comment:
Is this different than the Wasm for the runtime? If so, how is this generated?

Reply:
No, it is exactly the same. However, `validation_function` is the term that is being used in Polkadot and I wanted to stick to this.